### ISSUE TITLE
add CUB-200-2011 to CoOp, display error causes an error to run

### ISSUE THREAD

**Issue by KevinLi-167** [KaiyangZhou/Dassl.pytorch#65]
# The number of categories is displayed incorrectly

The cub-200-2011 is 200 categories, but shown as 201.

CoOp did not support the CUB dataset, so I took caltech101.py as a copy and rewrote it with reference to imagenet.

## 1. CUB200 dataset compare
CUB200, like Caltech101, does not divide the dataset by folder, but by the classification file.

CUB200 doesn't provide json like cal101 directly, but I constructed json in the same format.

## 2. Try to add CUB200 dataset to CoOp

By reading. the `train_test_split.txt`, CUB200 only has train and test datasets.

It's like ImageNet only has train and val datasets.

Therefore, some of the code in modifying CoOp/datasets/cub200.py is based on CoOp/datasets/imagenet.py.

To create a cub200.py, it also needs to be modified the 3 funcs from `OxfordPets`.

I can provide the source code for details.

## 3. classes of Caltech101
Specifically,

Caltech101 has 101 categories, and after ignoring 2 categories (`IGNORED = ["BACKGROUND_Google", "Faces_easy"]`),

the total is shown as 200:

    
    ---------  ----------
    Dataset Caltech101
    # classes  100
    # train_x  100
    # val  100
    # test 2,465
    ---------  ----------
    

This content is normal.

But I can't find where the classes are assigned in CoOp.

I don't think I need to change the `dassl` code。

## 4. classes of my CUB200
Part of the information after running  `bash scripts/coop/main.sh cub200 rn50_ep25 end 16 1 False`

    
    Loading trainer: CoOp
    Loading dataset: CUB200
    mkdir_if_missing is runnning
    Reading split from G:\CoOp_dir\CoOp\DATA\CUB_200_2011\CUB_split_byKL.json
    Creating a 1-shot dataset
    Saving preprocessed few-shot data to G:\CoOp_dir\CoOp\DATA\CUB_200_2011\split_fewshot\shot_1-seed_1.pkl
    Building transform_train
    + random resized crop (size=(224, 224), scale=(0.08, 1.0))
    + random flip
    + to torch tensor of range [0, 1]
    + normalization (mean=[0.48145466, 0.4578275, 0.40821073], std=[0.26862954, 0.26130258, 0.27577711])
    Building transform_test
    + resize the smaller edge to 224
    + 224x224 center crop
    + to torch tensor of range [0, 1]
    + normalization (mean=[0.48145466, 0.4578275, 0.40821073], std=[0.26862954, 0.26130258, 0.27577711])
    ---------  ------
    Dataset CUB200
    # classes  201
    # train_x  200
    # val  5,794
    # test 5,794
    ---------  ------
    

## 5. Others
I modified a lot of CoOp.files and I will provide if needed.

(Include the json corresponding to the CUB and the code that generated it)

I have successfully run the CoOp `train` and `interpret_prompt` code for caltech101.

About `\CoOp\configs\trainers\CoOp\rn50_ep25.yaml`,The default epoch is set to 50, but I set it to 25 for cal101.
  
DATALOADER.TRAIN_X. .BATCH_SIZE , This parameter is very interesting. I tried set 2~32 for CUB,Sometimes there are different errors .

I am Chinese, so if you want,  we can communicate in Chinese .

## 6. The main problem 
 - How to know if I'm loading the CUB correctly 
 - Why the number of classes is incorrect

**Comment by KevinLi-167** [comment]
The CUB200 dataset I generated, the category number is 1-200 instead of 0-199.

And after reading a func ` train, val, test = OxfordPets.read_split(self.split_path, self.image_dir)` that was not used in cal101.

I realized that it might be because my CUB.split.json' catyID index didn't start at 0.

So I tried Simply replace the category id of 200 with 0. Problem solved.

-------------------------------

To locate this problem, I looked for the content of the `print` in two projects (CoOp and Dassl).

At CoOp.train.py.main ,'build_trainer'
`trainer = build_trainer(cfg)`
In this sentence, some functions of Dassl are called

I found in `Dassl.pytorch\dassl\engine\trainer.py`,
`class SimpleTrainer(TrainerBase).__init__(self, cfg):`
Someone called his function and passed in the wrong cfg, 
and the content in the `cfg` identifies the number of my category IDs as 201
```
DataManager.show_dataset_summary:
    table.append(["# classes", f"{self.num_classes:,}"])
```

Although after a `len()` function, the output of my category length is 200.

-------------------------------
To conclusion, I did ignore that cal101's json starts with categoryId=0 which given by CoOp.
But the toolbox's method of identifying lengths is not very smart ( COOP calculates len instead of reading json).
